File: 0401-0500/0479.largest-palindrome-product.cpp
#include "common.h"

class Solution {
public:
    int largestPalindrome(int n) {
        // n位数*n位数，结果在2n-1 ~ 2n之间
        // 由于结果是回文数，需要枚举n位， 10^8太大了
        // 枚举n位吧
        int up = pow(10, n) - 1;
        int down = pow(10, n - 1);
        typedef long long LL;
        for (int h = 0; h < 2; ++h) {
            for (int v = up; v >= down; --v) {
                string left = to_string(v);
                string right(left.rbegin() + h, left.rend());
                LL p = stoll(left + right);
                for (LL k = up; k * k >= p; --k) {
                    if (p % k == 0) return p % 1337;
                }
            }
        }
        return -1;
    }
};

int main() {
  int n; cin >> n;
  cout << Solution().largestPalindrome(n) << endl;
  return 0;
}

File: 0401-0500/0479-largest-palindrome-product.py
class Solution:
    def largestPalindrome(self, n: int) -> int:
        if n == 1:
            return 9
        for i in range((10 ** n) - 1, 10 ** (n-1) - 1, -1):
            p = int(str(i) + str(i)[::-1])
            # 注意round(x + 0.5)向上取整, 可以写成整数形式：t * t >= p
            for t in range((10 ** n) - 1, max(10 ** (n-1), int(round(p ** 0.5) + 0.5)) - 1, -1):
                if p % t == 0:
                    return p % 1337

print(Solution().largestPalindrome(2))
#print(Solution().largestPalindrome(1))
